### PR TITLE
Check for thin_ls binary in path when using ThinPoolWatcher

### LINKED ISSUE #1308: Check for thin_ls in path and short-circuit if it is not installed
This issue is to add a check for `thin_ls` being present and avoid using it if it is not.  `NewThinPoolWatcher` should fail if the package is not installed.  

related: #1306

### PR DESCRIPTION
**Pull request by pmorie**
Resolves #1308

### CODE REVIEW

**Comment by pmorie** [comment]
@timstclair

File: container/docker/factory.go [hunk 1]
```diff
@@ -197,24 +197,38 @@ func Register(factory info.MachineInfoFactory, fsInfo fs.FsInfo, ignoreMetrics c
...
-		// thin_ls.
-		dockerThinPoolName, err := dockerutil.DockerThinPoolName(*dockerInfo)
-		if err != nil {
-			return fmt.Errorf("couldn't find device mapper thin pool name: %v", err)
-		}
-
-		dockerMetadataDevice, err := dockerutil.DockerMetadataDevice(*dockerInfo)
-		if err != nil {
-			return fmt.Errorf("couldn't determine devicemapper metadata device")
+		_, err := devicemapper.ThinLsBinaryPresent()
```
Comment (derekwaynecarr) [inline container/docker/factory.go]: assume kubelet will not have to check this thing?  wondering if we should also list it as a soft/hard dependency there

File: container/docker/factory.go [hunk 2]
```diff
@@ -197,24 +197,38 @@ func Register(factory info.MachineInfoFactory, fsInfo fs.FsInfo, ignoreMetrics c
...
+		if err == nil {
+			// If the storage drive is devicemapper, create and start a
```
Comment (ncdc) [inline container/docker/factory.go]: driver

File: container/docker/factory.go (hunk 1 above)
Comment (pmorie) [inline container/docker/factory.go]: There's an issue to track dependencies of kubelet, and I've called it out there.  we neeed to figure out the scheme in general for tracking and verifying stuff like that.

File: devicemapper/dmsetup_client.go [hunk 3]
```diff
@@ -21,19 +21,26 @@ import (
 	"github.com/golang/glog"
 )
 
-// DmsetupClient is a low-level client for interacting with devicemapper via
-// the dmsetup utility.
+// DmsetupClient is a low-level client for interacting with device mapper via
+// the `dmsetup` utility, which is provided by the `device-mapper` package.
```
Comment (vishh) [inline devicemapper/dmsetup_client.go]: `package` as in a go package or `rpm` or `deb`?

Comment (pmorie) [inline devicemapper/dmsetup_client.go]: RPM

Comment (vishh) [inline devicemapper/dmsetup_client.go]: Can you update the comments? Actually, having a note about the package in the global `Readme.md` might be super helpful. Not necessary for this PR though.

**Comment by vishh** [comment]
LGTM

File: devicemapper/dmsetup_client.go (hunk 3 above)
Comment (pmorie) [inline devicemapper/dmsetup_client.go]: I'll update in a follow-up

**Comment by dchen1107** [comment]
The cAdvisor pr builder job doesn't run for a couple of days. @fejta and I looked into why ...

**Comment by fejta** [comment]
ok to test

**Comment by pwittrock** [comment]
ok to test

**Comment by fejta** [comment]
@pwittrock if that doesn't work try directing the comment at k8s-bot....

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by fejta** [comment]
Looks like it is testing now! http://kubekins.dls.corp.google.com:8081/job/cadvisor-pull-build-test-e2e/646/

**Comment by dchen1107** [comment]
@pmorie Looks like the test failed due to your change:

16:18:35 --- FAIL: TestRefresh (0.00s)
16:18:35 	thin_pool_watcher_test.go:147: no existing reservation - ok: actual usage did not match expected usage: expected: 23456 got: 0
16:18:35 	thin_pool_watcher_test.go:147: existing reservation - ok: actual usage did not match expected usage: expected: 34567 got: 0
16:18:35 === RUN   TestCheckReservation
16:18:35 --- PASS: TestCheckReservation (0.00s)
16:18:35 FAIL
16:18:35 FAIL	github.com/google/cadvisor/devicemapper	0.052s

**Comment by pmorie** [comment]
@dchen1107 :( i will TAL

**Comment by pmorie** [comment]
@pwittrock @fejta @dchen1107 Looks like the builder had an infra-related problem -- what's the process to trigger a retest?  (I'm assuming the bot won't listen to me)

**Comment by fejta** [comment]
@pwittrock or another admin needs to make the test this comment he did above

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by pmorie** [comment]
same issue.

**Comment by pwittrock** [comment]
Looks like the test instances were deleted.  Will check around with what happened.

**Comment by pwittrock** [comment]
I am working on fixing the test environment.

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by pwittrock** [comment]
Got this failure.  Is this a red herring?

>
--- FAIL: TestDockerContainerById (12.70s)
	assertions.go:150: 
                        
	Location:	docker_test.go:63
		
	Error:		No error is expected but got request "http://e2e-cadvisor-rhel-7:8080/api/v1.3/docker/4928ebfc16223b2c44f0366da2b9b22753dbfd70e95d247e164e2bb0ef030265" failed with error: "failed to get Docker container \"4928ebfc16223b2c44f0366da2b9b22753dbfd70e95d247e164e2bb0ef030265\" with error: unable to find Docker container \"4928ebfc16223b2c44f0366da2b9b22753dbfd70e95d247e164e2bb0ef030265\""
		
	Messages:	Timed out waiting for container "4928ebfc16223b2c44f0366da2b9b22753dbfd70e95d247e164e2bb0ef030265" to be available in cAdvisor: request "http://e2e-cadvisor-rhel-7:8080/api/v1.3/docker/4928ebfc16223b2c44f0366da2b9b22753dbfd70e95d247e164e2bb0ef030265" failed with error: "failed to get Docker container \"4928ebfc16223b2c44f0366da2b9b22753dbfd70e95d247e164e2bb0ef030265\" with error: unable to find Docker container \"4928ebfc16223b2c44f0366da2b9b22753dbfd70e95d247e164e2bb0ef030265\""
		
FAIL
FAIL	github.com/google/cadvisor/integration/tests/api	87.088s

[https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/pull/1327/cadvisor-pull-build-test-e2e/652/build-log.txt](Full Logs)

**Comment by pwittrock** [comment]
@k8s-bot test this

**Comment by pwittrock** [comment]
Hm, looks like it was probably a flake then.

**Comment by dchen1107** [comment]
LGTM

**Pull request merged**